2022/day4/day4.rs

```diff
@@ -0,0 +1,57 @@
+///! [Day 4: Camp Cleanup](https://adventofcode.com/2022/day/4)
+use regex::Regex;
+use std::cmp::{max, min};
+
+struct Puzzle {
+    part1: i32,
+    part2: i32,
+}
+
+impl Puzzle {
+    fn new() -> Puzzle {
+        Puzzle { part1: 0, part2: 0 }
+    }
+
+    fn configure(&mut self, path: &str) {
+        let data = std::fs::read_to_string(path).unwrap();
+
+        let re = Regex::new(r"^(\d+)-(\d+),(\d+)-(\d+)").unwrap();
+
+        self.part1 = 0;
+        self.part2 = 0;
+
+        for line in data.split('\n').collect::<Vec<_>>() {
+            if let Some(m) = re.captures(line) {
+                let a = m[1].parse::<i32>().unwrap();
+                let b = m[2].parse::<i32>().unwrap();
+                let c = m[3].parse::<i32>().unwrap();
+                let d = m[4].parse::<i32>().unwrap();
+
+                if (a <= c && c <= d && d <= b) || (c <= a && a <= b && b <= d) {
```

Comment:
Heu, `c <= d` et `a <= b` c'est un peu une hypothèse du problème :stuck_out_tongue:

Reply:
Sauf si l'entrée est erronée, le programme et le compilateur n'en savent rien !

Reply:
Ah, donc on veut le faire à la mode rigoureux ?

Reply:
C'est surtout l'adaptation de l'expression Python  `a <= c <= d <= b` que je ne sais pas écrire en une fois en Rust.

Reply:
Ah, bonne question, je regarde mais pas sûr que ça se transcrive bien

Reply:
Impossible à faire en rust … pour des problèmes de compréhension du sens de l’expression

Reply:
Bah, j'ai l'impression qu'on est censé pourvoir faire `[a, c, d] < [c, d, b]` pour remplacer `a <= c && c <= d && d <= b` mais ça ne fait pas du tout ce que j'attendais. Pourtant sur Internet ils disent que c'est censé itérer en parallèle et vérifier... Je dois louper un truc, j'ai l'impression que ça ne vérifie que le premier élément :

```[2, 6, 8] <= [6, 8, 4] == true
[6, 2, 4] <= [2, 4, 8] == false
[2, 4, 5] <= [4, 5, 3] == true
[4, 2, 3] <= [2, 3, 5] == false
[5, 7, 9] <= [7, 9, 7] == true
[7, 5, 7] <= [5, 7, 9] == false
[2, 3, 7] <= [3, 7, 8] == true
[3, 2, 8] <= [2, 8, 7] == false
[6, 4, 6] <= [4, 6, 6] == false
[4, 6, 6] <= [6, 6, 6] == true
[2, 4, 8] <= [4, 8, 6] == true
[4, 2, 6] <= [2, 6, 8] == false

Reply:
C’est l’ordre lexicographique probablement